Add extension desktop UI

cryptotavares opened the pull request:
## Explanation
Adds UI code required for allowing the Extension to pair with the [MetaMask Desktop](https://github.com/MetaMask/metamask-desktop) app.

This PR includes:
* desktop pairing page -> this is the page where we will render an OTP to be introduced in the Desktop App.
* desktop error pages -> Extension error pages to be displayed when something goes wrong while connected to the Desktop App
* desktop enable button -> button to be displayed within settings->experimental page, that will allow enable/disable pairing the Extension with the Desktop App.
* desktop deep linking component -> allows opening up the Desktop App from the Extension if available and installed.
* adds desktop icon to header component when the Extension is running in dev mode and is paired with the Desktop app.
* register desktop listeners on UI launch -> this allows us to provide proper error handler for the edge cases where a user closes the Desktop App without going through the disable pairing process.
* disables ledger live control on the experimental advanced tab when paired with Desktop App

Besides the functional changes mentioned above, it is worth mentioning that we also extended the test render helper to allow proper testing of routes making use of the `useParam` hook.

## Screenshots/Screencaps
### Enable/Disable Button
![Screenshot 2023-02-14 at 17 02 49](https://user-images.githubusercontent.com/15957235/218808759-367e2f8a-cda9-40a1-8895-505f3798ea51.png)
![Screenshot 2023-02-14 at 17 08 21](https://user-images.githubusercontent.com/15957235/218808789-375e8dfb-1d33-4d8d-8626-5cabf6a13b59.png)
----------------------------
### Pairing page
![Screenshot 2023-02-14 at 17 07 51](https://user-images.githubusercontent.com/15957235/218808863-1ecfd6dd-ff57-443b-a26e-a29ecc8414f9.png)
----------------------------
### Error pages
![Screenshot 2023-02-14 at 17 09 30](https://user-images.githubusercontent.com/15957235/218808904-0114304e-605e-4e6b-a502-61a6a0f91376.png)
![Screenshot 2023-02-14 at 17 03 06](https://user-images.githubusercontent.com/15957235/218808960-4d007376-a1b6-4e38-862d-d766ff6bdc53.png)

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

## Manual Testing Steps
**Paring page:**
* Build the extension for desktop build type - `yarn start --build-type desktop`
* Navigate to settings->experimental
* Click on enable desktop button

**Error pages:**
* Build the extension for desktop build type - `yarn start --build-type desktop`
* Manually mutate the extension state to desktopEnabled set to true.
* Reload the extension
* Open the extension. It should try connect to the Desktop App. As there's none running, it will throw and display the error page.
<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
  - [x] How this problem was solved
  - [x] How reviewers can test my changes
- [x] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [ ] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

cryptotavares committed:
feat: add desktop enable button component

This component will be added
to the experimental page. Users
will then be able to initialize
a desktop connection

cryptotavares committed:
feat: add desktop pairing page

cryptotavares committed:
feat: add desktop deep-linking shared lib

cryptotavares committed:
test: add initial entries to render helper

Allow specifying initialEntries for
MemoryRouter. This change will allow
testing pages that use the useParam
hook.

cryptotavares committed:
feat: add desktop error page

Error page for any desktop pairing
related issue

cryptotavares committed:
feat: add desktop routes to route component

cryptotavares committed:
feat: add enable desktop button to experimental tab

cryptotavares committed:
feat: add desktop icon when paired in dev mode

cryptotavares committed:
feat: disable ledger live control when desktop enabled

cryptotavares committed:
feat: register desktop error actions on ui init

matthewwalsh0 commented on app/scripts/ui.js:
```diff
@@ -325,14 +338,21 @@ function initializeUi(activeTab, connectionStream, cb) {
 
   container.innerHTML = html;
 
+  registerDesktopErrorActions(backgroundConnection, browser);
```
Should we be code fencing the desktop specific UI code also? So functions like this and also moving the desktop components into a `desktop` folder so they'll be excluded from the other build types?

cryptotavares replied:
Yes I'll add the desktop code fencing everywhere... but we should not be moving the components to a specific desktop dir, because that would cause issues when we release under flask (as anything within `**/desktop/**` will only be available on desktop build type)

matthewwalsh0 commented on shared/lib/error-utils.js:
```diff
@@ -50,6 +54,22 @@ export async function getErrorHtml(errorKey, supportLink, metamaskState) {
+  if (isDesktopEnabled) {
+    let errorType = EXTENSION_ERROR_PAGE_TYPES.CRITICAL_ERROR;
+
+    if (err?.message.includes('No response from RPC')) {
```
Do we need to use the `t` method here in case it's an alternate language or is this error fixed?

cryptotavares replied:
Not really. This is not a UI error, but rather an error that gets thrown whenever the metaRPCClientFactory fails to get the state. Which when desktop is enabled we assume that it is due to a connection lost.

matthewwalsh0 commented on shared/lib/error-utils.js:
```diff
@@ -87,3 +107,57 @@ export async function getErrorHtml(errorKey, supportLink, metamaskState) {
+}
+
+export function downloadDesktopApp() {
+  global.platform.openTab({ url: 'https://metamask.io/' });
```
This could be a constant in the new `shared/constants/desktop.ts`.

cryptotavares replied:
We have an [open task](https://github.com/orgs/MetaMask/projects/37/views/4?filterQuery=link&pane=issue&itemId=20544786) to review all the desktop links. We can tackle it all within that task.

matthewwalsh0 commented:
We can remove the middleware handler too now, it's functional but unnecessary with the UI and it introduces more security risks.

`app/scripts/lib/rpc-method-middleware/handlers/desktop/enable-desktop.js`

cryptotavares committed:
fix: add missing code fencing

cryptotavares committed:
chore: remove enable desktop rpc middleware

Now that we are adding the UI
there's no need for this rpc middleware
(as it was used to test desktop background
code)

cryptotavares committed:
fix: display experimental tab for desktop

cryptotavares commented:
> Regarding the error screenshots, I've only ever seen the red cross in the horizontal centre, not sure how you got it on the left?

@matthewwalsh0 Yes you are right. There was a typo when converting from `ALIGN_ITEMS` to the new `AlignItems`. It is fixed now and the cross icon appears centered.
![Screenshot 2023-02-21 at 15 29 11](https://user-images.githubusercontent.com/15957235/220388812-fc142a38-9155-4cfb-a033-2e461aaec3c8.png)

jpuri reviewed (approved):
Great work 👍
